File: problems/LeetCode/top100/missingNumber.ts
function missingNumber(nums: number[]): number {
  let n = nums.length;
  let total = (n * (n + 1)) / 2;
  const sum = nums.reduce((acc, cur) => acc + cur, 0);

  return total - sum;
}

function missingNumber2(nums: number[]): number {
  let res = nums.length;

  for (let i = 0; i < nums.length; i++) {
    res += i - nums[i];
  }

  return res;
}

function missingNumber3(nums: number[]): number {
  let res = nums.length;
  for (let i = 0; i < nums.length; i++) {
    res ^= i ^ nums[i];
  }

  return res;
}
// Time: O(n)
// Space: O(1)

File: problems/LeetCode/NeetCode150/missingNumber.py
class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        s = set(nums)
        for i in range(len(nums) + 1):
            if i not in s:
                return i
            
class Solution:
    def missingNumber(self, nums: List[int]) -> int:

        res = len(nums)

        for i, num in enumerate(nums):
            res ^= i ^ num

        return res
    
class Solution:
    def missingNumber(self, nums: List[int]) -> int:

        n = len(nums)

        total = (1 + n) * n // 2

        return total - sum(nums)


# Time: O(n)
# Space: O(1)